lib/manifest/PlaylistGenerator.js

```diff
@@ -1,26 +1,340 @@
+};
+
+PlaylistGenerator.prototype.getDiagnostics = function (opts) {
+    var that = this;
+    return that.playlistImp.getDiagnostics(opts);
+};
+
+PlaylistGenerator.prototype.remove = function(){
+    var that = this;
+    return (that.stopPromise.promise ? that.stopPromise.promise : qResolved)
+        .then( function () {
+            return (that.persistPromise ? that.persistPromise : qResolved)
+        })
+        .then(function () {
+            return qio.remove(that.playlistPath);
+        });
+};
+
+
+var persistPlayList = function persistPlayList(force){
+    var that = this;
+
+    if(that.isDirty || force) {
+
+        if(that.persistPromise){
+            return that.persistPromise
+                .then( function(){
+                    return persistPlayList.call(that,force);
+                });
+        }
+
+        if(that.playlistImp.durations.length > 0) {
+            var minMaxFirstClip = that.playlistImp.recalculateOffsetsAndDuration(0),
+                minMaxLastClip = that.playlistImp.durations.length > 1 ? that.playlistImp.recalculateOffsetsAndDuration(that.playlistImp.durations.length - 1) : minMaxFirstClip;
+
+            if(!(minMaxFirstClip && minMaxLastClip)){
+                return qResolved;
+            }
+
+            // only update manifest if all downloaders have contributed i.e. playback range changed
+            if (!force && that.minMax && that.minMax.min === minMaxFirstClip.min && that.minMax.max === minMaxLastClip.max) {
+                return qResolved;
+            }
+
+            if(that.playlistImp.segmentBaseTime === 0){
+                that.playlistImp.segmentBaseTime = that.playlistImp.firstClipTime;
+            }
+
+            that.minMax = {min: minMaxFirstClip.min, max: minMaxLastClip.max};
+
+            if(this.discontinuity) {
+                var segmentDuration  = config.get('vodPackagerSegmentDuration') || 10000;
+                that.playlistImp.initialSegmentIndex = Math.floor((that.playlistImp.firstClipTime - that.playlistImp.segmentBaseTime) / segmentDuration);
+            }
+
+        } else {
+            delete that.minMax;
+        }
+        that.isDirty = false;
+        try {
+            var str = JSON.stringify(that.playlistImp);
+            return that.persistPromise = fsUtils.writeFileAtomically(that.playlistPath,JSON.stringify(that.playlistImp));
+        } catch(err){
+            return Q.resolve(err);
+        }
+    }
+    return qResolved;
+};
+
+var finaizeStart = function (){
+    var that = this;
+
+    if (!that.stopPromise) {
+        that.stopPromise = Q.defer();
+    }
+
+    that.playlistImp.addListener(handleEvents.bind(that));
+
+    that.startPromise.resolve();
+};
+
+PlaylistGenerator.prototype.start = function start() {
+    var that = this;
+
+    if(that.startPromise)
+        return that.startPromise;
+
+    try {
+        that.startPromise = Q.defer();
+        fs.exists(that.playlistPath,function(ok) {
```

Comment:
rather if you use promises here, it's one time thing, and it's better to be consist